src/Raven.Server/Documents/TimeSeries/TimeSeriesStorage.cs

```diff
@@ -374,96 +375,131 @@ bool TryRemoveRange(ref TableValueReader reader, out Table.TableValueHolder next
                             {
                                 var current = baseline.AddMilliseconds(ts);
 
-                                if (current >= from && current <= to)
+                                if (current > to && segmentChanged == false)
+                                    return false; // we can exit here earlier
+
+                                if (current >= from && current <= to && 
+                                    status == TimeSeriesValuesSegment.Live)
                                 {
                                     holder.AddNewValue(current, values, tag.AsSpan(), ref newSegment, TimeSeriesValuesSegment.Dead);
                                     removeOccured = true;
+                                    segmentChanged = true;
                                     continue;
                                 }
 
                                 holder.AddExistingValue(current, values, tag.AsSpan(), ref newSegment, status);
                             }
                         }
 
-                        holder.AppendExistingSegment(newSegment);
-                        changeVector = holder.ChangeVector;
+                        if (segmentChanged)
+                        {
+                            holder.AppendExistingSegment(newSegment);
+                            changeVector = holder.ChangeVector;
+                        }
+
                         return end < to;
                     }
                 }
 
-                Table.TableValueHolder TryGetNextSegment(DateTime baseline)
+                DateTime? TryGetNextBaseline()
                 {
                     var offset = slicer.TimeSeriesKeySlice.Size - sizeof(long);
                     *(long*)(slicer.TimeSeriesKeySlice.Content.Ptr + offset) = Bits.SwapBytes(baseline.Ticks / 10_000);
 
                     foreach (var (_, tvh) in table.SeekByPrimaryKeyPrefix(slicer.TimeSeriesPrefixSlice, slicer.TimeSeriesKeySlice, 0))
                     {
-                        return tvh;
+                        return GetBaseline(tvh.Reader);
                     }
 
                     return null;
                 }
 
-                void RemoveTimeSeriesNameFromMetadata(DocumentsOperationContext ctx, string docId, string tsName)
+                void Notify()
                 {
-                    var doc = _documentDatabase.DocumentsStorage.Get(ctx, docId);
-                    if (doc == null)
-                        return;
+                    context.Transaction.AddAfterCommitNotification(new TimeSeriesChange
+                    {
+                        ChangeVector = changeVector, // change-vector can be null, if we delete an entire segments
```

Comment:
Why? Shouldn't we still generate a change vector for the deletion?

Reply:
Since the optimization was removed, so it that comment.